### PR TITLE
Modernize Scancode Implementation

### PR DESCRIPTION
**Pull request by ChrisThrasher**
## Description

This addresses some review comments from #2372. Read this PR on a commit-by-commit basis to understand each change being made. We can squash it later if we want.

### CODE REVIEW

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Review by Bromeon — COMMENTED** [review]
Regarding all the `static_cast<std::size_t>(code)`, what do you think [about overloading `operator+`](https://stackoverflow.com/a/42198760/) so that this becomes `+code`?

Just asking because the number of `static_cast`s in SFML is _really_ becoming a readability burden.

File: include/SFML/Window/Keyboard.hpp [hunk 1]
```diff
@@ -174,183 +164,166 @@ enum Key
...
+    LaunchMediaSelect,  //!< Keyboard Launch Media Select key
 
-#if defined(__clang__)
-#pragma clang diagnostic pop
-#endif
+    ScancodeCount //!< Keep last -- the total number of scancodes
 };
 
-using Scancode = Scan::Scancode;
+using Scancode = Scan;
```
[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply include/SFML/Window/Keyboard.hpp:326]: Screenshot 2023-01-23 at 11 06 12 AM

Yes we do. Many places still refer to a type named `Scancode`. This one type has two names due to its history as a plain enum inside a struct which needed to have different names. Hopefully we can clean this up and pick only one name but I wasn't sure how to handle that in this PR so I left the type alias in place.

File: src/SFML/Window/Android/WindowImplAndroid.cpp [hunk 2]
```diff
@@ -647,7 +647,7 @@ Keyboard::Key WindowImplAndroid::androidKeyToSF(std::int32_t key)
         case AKEYCODE_RIGHT_BRACKET:      return Keyboard::RBracket;
         case AKEYCODE_BACKSLASH:          return Keyboard::Backslash;
         case AKEYCODE_SEMICOLON:          return Keyboard::Semicolon;
-        case AKEYCODE_APOSTROPHE:         return Keyboard::Quote;
+        case AKEYCODE_APOSTROPHE:         return Keyboard::Apostrophe;
```
[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply src/SFML/Window/Android/WindowImplAndroid.cpp:650]: It's not a bug. It's just that `Keyboard::Quote` was deprecated in 2.6.0 and removed in 3.0 so I had to switch to the non-deprecated version.

**Comment by ChrisThrasher** [comment]
> what do you think [about overloading operator+](https://stackoverflow.com/a/42198760/) so that this becomes +code?

I used this syntax before to convert `uint8_t`s to `int`s for the sake of printing and it was rejected on readability grounds so while I'm open to this idea, I worry others won't like it.

> Just asking because the number of static_casts in SFML is really becoming a readability burden.

Agreed that it's a major eye sore. I have to bring up that using functional casts is a really easy way to cut down on the eyesore of `static_cast<std::size_t>(...)` but I know that's not a popular opinion.

File: src/SFML/Window/Unix/KeyboardImpl.cpp [hunk 3]
```diff
@@ -279,172 +279,172 @@ std::map<std::string, sf::Keyboard::Scancode> getNameScancodeMap()
 {
     std::map<std::string, sf::Keyboard::Scancode> mapping;
```
Comment (kimci86) [inline src/SFML/Window/Unix/KeyboardImpl.cpp:280]: I believe unordered_map is preferred.
```suggestion
    std::unordered_map<std::string, sf::Keyboard::Scancode> mapping;
```

File: src/SFML/Window/Unix/KeySymToKeyMapping.hpp [hunk 4]
```diff
@@ -1,7 +1,7 @@
 ////////////////////////////////////////////////////////////
 //
 // SFML - Simple and Fast Multimedia Library
-// Copyright (C) 2007-2021 Laurent Gomila (laurent@sfml-dev.org)
+// Copyright (C) 2007-2023 Laurent Gomila (laurent@sfml-dev.org)
```
Comment (kimci86) [inline src/SFML/Window/Unix/KeySymToKeyMapping.hpp:4]: Not to forget that this can be updated on 2.6.x too.

**Review by kimci86 — COMMENTED** [review]
Do we want to make Keyboard::Key a scoped enumeration too? Maybe out of scope for this PR.

**Comment by ChrisThrasher** [comment]
> Do we want to make Keyboard::Key a scoped enumeration too? Maybe out of scope for this PR

I think we do but idk what API implications that would have so I was afraid to add that to this PR. I want the key code and scan code interfaces to be as symmetric as possible and currently that is not the case.

Also, are they called "keys and scans" or "keys and scan codes" or "key codes and scan codes"? Idk how exactly we're supposed to refer to these things.

**Comment by Bromeon** [comment]
Maybe the naming should indeed be deferred to another PR, I remember there was already quite some discussion once we introduced it.

File: src/SFML/Window/Unix/KeySymToKeyMapping.hpp (hunk 4 above)
Comment (eXpl0it3r) [reply src/SFML/Window/Unix/KeySymToKeyMapping.hpp:4]: Done with #2376

**Review by kimci86 — APPROVED** [review]

File: include/SFML/Window/Keyboard.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (eXpl0it3r) [reply include/SFML/Window/Keyboard.hpp:326]: Btw, [here's the discussion thread](https://github.com/SFML/SFML/pull/1235#discussion_r642070280) with some of @Bromeon thoughts on the SFML 2 and SFML 3 design.

There's still time to align on either side.

File: src/SFML/Window/Android/WindowImplAndroid.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (eXpl0it3r) [reply src/SFML/Window/Android/WindowImplAndroid.cpp:650]: I guess, we should switch it also in 2.6, so we're not returning a deprecated value, even if it matches the same numerical value and can't really be distinguished. More for the prettiness of things 😄

**Review by eXpl0it3r — APPROVED** [review]

**Pull request merged**